Add a layout test for volatile front buffer (forcing a sync GPUP swap).

mattwoodrow opened the pull request:
#### 247a813a16d6f0364d32806a035e0a4e5d52ac32
<pre>
Add a layout test for volatile front buffer (forcing a sync GPUP swap).
<a href="https://bugs.webkit.org/show_bug.cgi?id=267749">https://bugs.webkit.org/show_bug.cgi?id=267749</a>
&lt;<a href="https://rdar.apple.com/121232490">rdar://121232490</a>&gt;

Reviewed by Simon Fraser.

Adds a new &apos;markFrontBufferVolatileForTesting&apos; function to Internals, which sets
the volatile state on the current front buffer of the Element&apos;s layer.

Adds a test that uses this, and would catch the failure from bug 267300.

* LayoutTests/compositing/repaint/needs-no-display-volatile-expected.html: Added.
* LayoutTests/compositing/repaint/needs-no-display-volatile.html: Added.
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::markFrontBufferVolatileForTesting):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::markFrontBufferVolatileForTesting):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::markFrontBufferVolatileForTesting):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::markFrontBufferVolatile):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::markFrontBufferVolatileForTesting):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerBackingStoreCollection::markFrontBufferVolatileForTesting):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::prepareToDisplay):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::markFrontBufferVolatileForTesting):

Canonical link: <a href="https://commits.webkit.org/273814@main">https://commits.webkit.org/273814@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5bdc06a763bff917271ea8026371169246b0dab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36744 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15680 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38982 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39387 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37989 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12792 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13212 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32473 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11561 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32747 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40636 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33279 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33091 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37492 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11846 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9665 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35612 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32425 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/4761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12742 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
```diff
@@ -259,6 +259,23 @@
+void RemoteLayerBackingStoreCollection::markFrontBufferVolatileForTesting(RemoteLayerBackingStore& backingStore)
+{
+    if (auto* remoteBackingStore = dynamicDowncast<RemoteLayerWithRemoteRenderingBackingStore>(&backingStore)) {
+
```
Blank line here.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
```diff
@@ -259,6 +259,23 @@
+    if (auto* remoteBackingStore = dynamicDowncast<RemoteLayerWithRemoteRenderingBackingStore>(&backingStore)) {
+
+        auto bufferSet = remoteBackingStore->protectedBufferSet();
+        if (bufferSet) {
```
`if (auto bufferSet = remoteBackingStore->protectedBufferSet())`

cdumez replied:
We normally prefer:
```
if (RefPtr bufferSet = remoteBackingStore->bufferSet())
```

The `protectedFoo()` functions are to avoid local variables.

cdumez replied:
Please don't hide smart pointers behind `auto`. As we're adopting smart pointers, it is very convenient to see right away that we have a `RefPtr` here.

cdumez commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -5078,6 +5078,12 @@ void GraphicsLayerCA::purgeBackBufferForTesting()
 
+void GraphicsLayerCA::markFrontBufferVolatileForTesting()
+{
+    if (primaryLayer())
```
Please should probably use a RefPtr here per our guidelines:
```
if (RefPtr layer = primaryLayer())
    layer->markFrontBufferVolatileForTesting();
```

cdumez commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5803,6 +5803,12 @@ void RenderLayer::purgeBackBufferForTesting()
 
+void RenderLayer::markFrontBufferVolatileForTesting()
+{
+    if (backing())
```
Ideally we'd use a CheckedPtr here since the backing is not ref counted.

cdumez commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -4326,4 +4326,10 @@ void RenderLayerBacking::purgeBackBufferForTesting()
 
+void RenderLayerBacking::markFrontBufferVolatileForTesting()
+{
+    if (m_graphicsLayer)
```
You need a local smart pointer, per our guidelines:
```
if (RefPtr graphicsLayer = m_graphicsLayer)
    graphicsLayer->markFrontBufferVolatileForTesting();
```

cdumez commented on Source/WebCore/rendering/RenderLayerBacking.h:
```diff
@@ -288,10 +288,12 @@ class RenderLayerBacking final : public GraphicsLayerClient {
     WEBCORE_EXPORT void setIsTrackingDisplayListReplay(bool);
     WEBCORE_EXPORT String replayDisplayListAsText(OptionSet<DisplayList::AsTextFlag>) const;
 
+
```
Extra blank line here.

cdumez commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -1394,6 +1394,12 @@ void Internals::purgeBackBuffer(Element& element)
+void Internals::markFrontBufferVolatile(Element& element)
+{
+    if (element.renderer() && element.renderer()->enclosingLayer())
+        element.renderer()->enclosingLayer()->markFrontBufferVolatileForTesting();
```
enclosingLayer should be in a CheckedPtr before calling the member function.

cdumez commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -600,6 +600,12 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
 
+void RemoteLayerBackingStore::markFrontBufferVolatileForTesting()
+{
+    if (auto* collection = backingStoreCollection())
```
Ideally would use a CheckedPtr since not ref counted.

cdumez commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
```diff
@@ -254,6 +254,23 @@
+            sendMarkBuffersVolatile(WTFMove(identifiers), [](bool) { });
+        }
+    } else {
+        auto& inProcessBackingStore = downcast<RemoteLayerWithInProcessRenderingBackingStore>(backingStore);
```
CheckedRef.

cdumez commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
```diff
@@ -1101,5 +1101,11 @@ static inline bool isEquivalentLayer(const PlatformCALayer* layer, const std::op
 
+void PlatformCALayerRemote::markFrontBufferVolatileForTesting()
+{
+    if (m_properties.backingStoreOrProperties.store)
```
Should put the store in a CheckedPtr local before calling the member function on it (since not ref counted).

mattwoodrow committed:
Add a layout test for volatile front buffer (forcing a sync GPUP swap).
https://bugs.webkit.org/show_bug.cgi?id=267749
<rdar://121232490>

Reviewed by Simon Fraser.

Adds a new 'markFrontBufferVolatileForTesting' function to Internals, which sets
the volatile state on the current front buffer of the Element's layer.

Adds a test that uses this, and would catch the failure from bug 267300.

* LayoutTests/compositing/repaint/needs-no-display-volatile-expected.html: Added.
* LayoutTests/compositing/repaint/needs-no-display-volatile.html: Added.
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::markFrontBufferVolatileForTesting):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::markFrontBufferVolatileForTesting):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::markFrontBufferVolatileForTesting):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::markFrontBufferVolatile):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::ensureBufferForDisplay):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::markFrontBufferVolatileForTesting):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::prepareBackingStoresForDisplay):
(WebKit::RemoteLayerBackingStoreCollection::markFrontBufferVolatileForTesting):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::prepareToDisplay):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::markFrontBufferVolatileForTesting):

Canonical link: https://commits.webkit.org/273814@main
